core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -0,0 +1,576 @@
+                || aggregation.getMask().isPresent()
+                || aggregation.getOrderingScheme().isPresent()) {
+            // aggregation must be a basic aggregation
+            return Optional.empty();
+        }
+
+        String name = aggregation.getResolvedFunction().getSignature().getName();
+        if (!ALLOWED_FUNCTIONS.contains(name)) {
+            // only cumulative aggregations (e.g that can be split into aggregation of aggregations) are supported
+            return Optional.empty();
+        }
+
+        Symbol projectionSymbol = Symbol.from(aggregation.getArguments().get(0));
+        Expression projection = projectNode.getAssignments().get(projectionSymbol);
+        Expression unwrappedProjection;
+        // unwrap top-level cast
+        if (projection instanceof Cast) {
+            unwrappedProjection = ((Cast) projection).getExpression();
+        }
+        else {
+            unwrappedProjection = projection;
+        }
+
+        if (!(unwrappedProjection instanceof SearchedCaseExpression)) {
+            return Optional.empty();
+        }
+
+        SearchedCaseExpression caseExpression = (SearchedCaseExpression) unwrappedProjection;
+        if (caseExpression.getWhenClauses().size() != 1) {
+            return Optional.empty();
+        }
+
+        if (caseExpression.getDefaultValue().isPresent()) {
+            Type defaultType = getType(context, caseExpression.getDefaultValue().get());
+            Object defaultValue = optimizeExpression(caseExpression.getDefaultValue().get(), context);
+            if (defaultValue != null) {
+                if (!name.equals("sum")) {
+                    return Optional.empty();
+                }
+
+                // sum aggregation is only supported if default value is null or 0, otherwise it wouldn't be cumulative
+                if (defaultType instanceof BigintType
+                        || defaultType == INTEGER
+                        || defaultType == SMALLINT
+                        || defaultType == TINYINT
+                        || defaultType == DOUBLE
+                        || defaultType == REAL
+                        || defaultType instanceof DecimalType) {
+                    if (!defaultValue.equals(0L) && !defaultValue.equals(0.0d) && !defaultValue.equals(Int128.ZERO)) {
+                        return Optional.empty();
+                    }
+                }
+                else {
+                    return Optional.empty();
+                }
+            }
+        }
+
+        Type aggregationType = aggregation.getResolvedFunction().getSignature().getReturnType();
+        ResolvedFunction cumulativeFunction;
+        try {
+            cumulativeFunction = plannerContext.getMetadata().resolveFunction(context.getSession(), QualifiedName.of(name), fromTypes(aggregationType));
+        }
+        catch (TrinoException e) {
+            // there is no cumulative aggregation
+            return Optional.empty();
```

Comment:
Can this happen?
I think the `name` corresponds to an agg function used in the plan, so it should be expected to exist.